src/query_graph/graph_path.rs

```diff
@@ -183,3 +184,22 @@ pub(crate) struct ClosedBranch(Vec<Arc<ClosedPath>>);
 /// A list of the options generated during query planning for a specific "open branch", which is a
 /// partial/open path in a GraphQL operation (i.e. one that does not end in a leaf field).
 pub(crate) struct OpenBranch(Vec<SimultaneousPathsWithLazyIndirectPaths>);
+
+impl TryFrom<&'_ OpPath> for Vec<QueryPathElement> {
+    type Error = FederationError;
+
+    fn try_from(value: &'_ OpPath) -> Result<Self, Self::Error> {
+        value
+            .0
+            .iter()
+            .map(|path_element| {
+                Ok(match path_element.as_ref() {
+                    OpPathElement::Field(field) => QueryPathElement::Field(field.try_into()?),
+                    OpPathElement::InlineFragment(inline) => {
+                        QueryPathElement::InlineFragment(inline.try_into()?)
+                    }
+                })
+            })
+            .collect()
+    }
```

Comment:
Something cool but maybe not obvious: [`collect()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect) can turn an iterator of results into a single result (stopping at the first error)